core/error_context/query_public_json.hxx

```diff
@@ -0,0 +1,87 @@
+/* -*- Mode: C++; tab-width: 4; c-basic-offset: 4; indent-tabs-mode: nil -*- */
+/*
+ *   Copyright 2024. Couchbase, Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+#pragma once
+
+#include <couchbase/fmt/retry_reason.hxx>
+#include <couchbase/query_error_context.hxx>
+
+#include <fmt/format.h>
+#include <tao/json/forward.hpp>
+
+namespace tao::json
+{
+template<>
+struct traits<couchbase::query_error_context> {
+    template<template<typename...> class Traits>
+    static void assign(tao::json::basic_value<Traits>& v, const couchbase::query_error_context& ctx)
+    {
+        if (ctx.last_dispatched_to()) {
+            v["last_dispatched_to"] = ctx.last_dispatched_to().value();
+        }
+        if (ctx.last_dispatched_from()) {
+            v["last_dispatched_from"] = ctx.last_dispatched_from().value();
+        }
+        if (ctx.retry_attempts()) {
+            v["retry_attempts"] = ctx.retry_attempts();
+        }
+        if (!ctx.retry_reasons().empty()) {
+            std::vector<tao::json::basic_value<Traits>> reasons{};
+            for (couchbase::retry_reason r : ctx.retry_reasons()) {
+                reasons.emplace_back(fmt::format("{}", r));
+            }
+            v["retry_reasons"] = reasons;
+        }
```

Comment:
Could it be better to always include the retry attempts & reasons fields in the JSON even when they're zero/empty?